stdlib/mexpr/cfa.mc

```diff
@@ -587,249 +676,282 @@ end
-  | CSym2hash _ -> []
+  sem propagateConstraintConst res args graph info =
+  | CSymb _ -> graph
+  | CGensym _ -> graph
+  | CSym2hash _ -> graph
 end
 
 lang CmpSymbCFA = CFA + ConstCFA + CmpSymbAst
-  sem generateConstraintsConst (info: Info) =
-  | CEqsym _ -> []
+  sem propagateConstraintConst res args graph info =
+  | CEqsym _ -> graph
 end
 
--- TODO(dlunde,2021-11-11): Add flow constraints for sequence operations?
-lang SeqOpCFA = CFA + ConstCFA + SeqOpAst
-  sem generateConstraintsConst (info: Info) =
-  -- | CSet _ -> []
-  -- | CGet _ -> []
-  -- | CCons _ -> []
-  -- | CSnoc _ -> []
-  -- | CConcat _ -> []
-  -- | CLength _ -> []
-  -- | CReverse _ -> []
-  -- | CHead _ -> []
-  -- | CTail _ -> []
-  -- | CNull _ -> []
-  -- | CMap _ -> []
-  -- | CMapi _ -> []
-  -- | CIter _ -> []
-  -- | CIteri _ -> []
-  -- | CFoldl _ -> []
-  -- | CFoldr _ -> []
-  -- | CCreate _ -> []
-  -- | CCreateList _ -> []
-  -- | CCreateRope _ -> []
-  -- | CIsList _ -> []
-  -- | CIsRope _ -> []
-  -- | CSplitAt _ -> []
-  -- | CSubsequence _ -> []
+lang SeqOpCFA = CFA + ConstCFA + AppCFA + SeqCFA + SeqOpAst
+  sem propagateConstraintConst res args graph info =
+  | CSet _ ->
+    utest length args with 3 in
+    let seq = get args 0 in
+    let val = get args 2 in
+    initConstraint graph (CstrSeqUnion {lhs = seq, rhs = val, res = res})
+  | CGet _ ->
+    utest length args with 2 in
+    initConstraint graph (CstrSeq {lhs = head args, rhs = res})
+  | CCons _ ->
+    utest length args with 2 in
+    let val = get args 0 in
+    let seq = get args 1 in
+    initConstraint graph (CstrSeqUnion {lhs = seq, rhs = val, res = res})
+  | CSnoc _ ->
+    utest length args with 2 in
+    let seq = get args 0 in
+    let val = get args 1 in
+    initConstraint graph (CstrSeqUnion {lhs = seq, rhs = val, res = res})
+  | CConcat _ ->
+    utest length args with 2 in
+    let graph = initConstraint graph (CstrDirect {lhs = head args, rhs = res}) in
+    initConstraint graph (CstrDirect {lhs = get args 1, rhs = res})
+  | CLength _ ->
+    utest length args with 1 in
+    graph
+  | CReverse _ ->
+    utest length args with 1 in
+    initConstraint graph (CstrDirect {lhs = head args, rhs = res})
+  | CHead _ ->
+    utest length args with 1 in
+    initConstraint graph (CstrSeq {lhs = head args, rhs = res})
+  | CTail _ ->
+    utest length args with 1 in
+    initConstraint graph (CstrDirect {lhs = head args, rhs = res})
+  | CNull _ ->
+    utest length args with 1 in
+    graph
+  | CIsList _ ->
+    utest length args with 1 in
+    graph
+  | CIsRope _ ->
+    utest length args with 1 in
+    graph
+  | CSubsequence _ ->
+    utest length args with 3 in
+    initConstraint graph (CstrDirect {lhs = head args, rhs = res})
+  | ( CMap _
+    | CMapi _
+    | CIter _
+    | CIteri _
+    | CFoldl _
+    | CFoldr _
+    | CCreate _
+    | CCreateList _
+    | CCreateRope _
+    | CSplitAt _
+    ) -> infoErrorExit info "Sequence intrinsic not supported in CFA"
```

Comment:
Looks good! But, I think that we should maybe define `CstrSeq` and `CstrSeqUnion` in this fragment instead, as they are only used here. What do you think?

Reply:
Good idea, I'll move them.